launcher/ui/pages/instance/McClient.cpp

```diff
@@ -0,0 +1,166 @@
+#include <QObject>
+#include <QTcpSocket>
+#include <QJsonDocument>
+#include <QJsonObject>
+#include <qtconcurrentrun.h>
+
+#include <Exception.h>
+#include "McClient.h"
+#include "Json.h"
+
+// 7 first bits 
+#define SEGMENT_BITS 0x7F
+// last bit
+#define CONTINUE_BIT 0x80
+
+McClient::McClient(QObject *parent, QString domain, QString ip, short port): QObject(parent), m_domain(domain), m_ip(ip), m_port(port) {}
+
+QJsonObject McClient::getStatusDataBlocking() {
+    qDebug() << "Connecting to socket..";
+    m_socket.connectToHost(m_ip, m_port);
+
+    if (!m_socket.waitForConnected()) {
+        throw Exception("Failed to connect to socket");
+    }
+    qDebug() << "Connected to socket successfully";
+    sendRequest();
+
+    if (!m_socket.waitForReadyRead()) {
+        throw Exception("Socket didn't send anything to read");
+    }
+    return readResponse();
+}
+
+QFuture<int> McClient::getOnlinePlayers() {
+    return QtConcurrent::run([this]() {
+        try {
+            auto status = getStatusDataBlocking();
+            auto players = Json::requireObject(status, "players");
+            return Json::requireInteger(players, "online");
+        } catch (const Exception &e) {
+            qDebug() << "Error: " << e.what();
+            return -1;
+        }
+    });
```

Comment:
It feels a bit pointless to use a thread when it is just waiting doing nothing while another thread does the actual work...

Also, this method is not recommended on Windows